pkg/pub_package_reader/lib/src/package_dir.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'tar_utils.dart';
+
+/// Helper methods to access package content.
+abstract class PackageDir {
+  /// List all the file names in the directory.
+  Future<List<String>> listFileNames();
+
+  /// Whether the file on the given relative [path] exists.
+  Future<bool> exists(String path);
+
+  /// Read the content of the file on the given relative [path].
+  Stream<List<int>> read(String path);
```

Comment:
What happens if the `path` doesn't exist?

Reply:
Can we just make this `Future<List<int>>` for simplicity?

Reply:
Should it also have a `maxLength` property to limit the memory use. Not sure if cleverly compressed archives are still a thing...